wwwcli: Adds human readable settings edituser

s-ben opened the pull request:
This commit adds human readable input settings to the edituser
command. Help command text is also updated to reflect new
human readable options. ( Closes #642 )

s-ben commented:
Presumably we do this by just summing the EmailNotifications variable instead of directly assigning it like we do here?
https://github.com/decred/politeia/blob/ae149f59921240a657f80b04776f4ddf646c37ff/politeiawww/user.go#L109

This is making me realize though that a) the user can only add notifications right now, not remove them (unless they pass negative integers to the edituser command, which is not very intuitive and not documented in the help currently), and b) there's no error handling on the passed integer. The user can currently pass in any integer, which could turn on or off random notifications.

lukebp commented:
> a) the user can only add notifications right now, not remove them (unless they pass negative integers to the edituser command, which is not very intuitive and not documented in the help currently)

They can remove them by simply passing up a new bit field with the appropriate bits flipped. 

> b) there's no error handling on the passed integer. The user can currently pass in any integer, which could turn on or off random notifications.

That was intentional.

Both of these commands should result in the same outcome.
```
$ politeiawwwcli edituser emailnotifications="proposalvetted,votingstarted,commentonproposal"
{
  "emailnotifications": 148,
}

$ politeiawwcli edituser emailnotifications="148"
{
  "emailnotifications": 148,
}
```

There should also be a human readable option for setting all notification bits to zero.  A `none` or something similar.

s-ben commented:
@lukebp , have implemented the ability to set multiple notifications. 

Note: per convention with other commands, the passed in variable now needs the `--` flag. So your example above:
```politeiawwwcli edituser emailnotifications="proposalvetted,votingstarted,commentonproposal"```

is now:
```politeiawwwcli edituser --emailnotifications="proposalvetted,votingstarted,commentonproposal"```

Still need to update the help text and rebase, but figured I'd hold off on that for now in case we still need more changes.

lukebp commented on politeiawww/cmd/politeiawwwcli/commands/edituser.go:
```diff
@@ -32,17 +37,60 @@ Response:
 
 func (cmd *EditUserCmd) Execute(args []string) error {
+
+	EmailNotifs := map[string]v1.EmailNotificationT{
```
This email notifications list should reference the actual notification variables defined in the api instead of hard coding in the values.

Example:
```
"userproposalchange": v1.NotificationEmailMyProposalStatusChange,
```

Also, I would avoid capitalizing local variables like `EmailNotifs`.  Exported variables in go are capitalized, so capitalizing a non-exported variable might be seen as confusing.  `emailNotifs` would be better.

s-ben replied:
Done. FYI, we also hard code values in `manageuser.go` (the function I was modeling this on). Should I also update `manageuser` to reference variables in v1? If so, do it in this PR, or open another?

lukebp replied:
You can include it in this PR.

lukebp commented on politeiawww/cmd/politeiawwwcli/commands/edituser.go:
```diff
@@ -32,17 +37,60 @@ Response:
+		// Human readable action code found
+		notif = a
+
+	} else if a, ok := EmailNotifs[strings.Split(cmd.Args.NotifType, ",")[0]]; ok {
```
This is kinda hard to read.  Do you think something like this would read better?

```
else if strings.Contains(cmd.Args.NotifType, ",") {
```

s-ben replied:
This does read better. Using this. Here I was again following `manageuser.go`, which uses this syntax. I did a quick search for this pattern in the rest of the codebase, couldn't find any other uses of it. It looks like `manageuser.go` has a few "style guide" sins, if those exist here.

lukebp replied:
Manage user is a bit different because a list isn't being passed in so checking for a comma doesn't make sense.  

As far as style goes, most of it comes from either [Effective Go](https://golang.org/doc/effective_go.html) or [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments).  The rest is usually personal preference.  [Practical Go](https://dave.cheney.net/practical-go/presentations/qcon-china.html#_comments) is also a good resource.

lukebp commented on politeiawww/cmd/politeiawwwcli/commands/edituser.go:
```diff
@@ -32,17 +37,60 @@ Response:
+		}
+
+	} else {
+		return fmt.Errorf("Invalid edituser option. Type 'help edituser' for list of valid options")
```
80 column limit.  Split these error messages into two lines.

s-ben replied:
Done.

lukebp commented on politeiawww/cmd/politeiawwwcli/commands/edituser.go:
```diff
@@ -32,17 +37,60 @@ Response:
+			if !ok {
+				return fmt.Errorf("Invalid edituser option. Type 'help edituser' for list of valid options")
+			}
+			notif += a
```
I would probably use a bitwise operator here to make it clear that you're working with bit flags.  This works too though.  Your call.

```
notif |= a
```

s-ben replied:
Makes sense. Using bitwise operator might make the code more readable.

lukebp commented on politeiawww/cmd/politeiawwwcli/commands/edituser.go:
```diff
@@ -32,17 +37,60 @@ Response:
+		notif = a
+		// Parse list of strings and calculate associated integer
+		s := strings.Split(cmd.Args.NotifType, ",")
+		for i := 1; i < len(s); i++ {
```
Would prefer range.

```
for _, v := range s {

}
```

s-ben replied:
Testing this I actually found a bug:/  In my old for loop, I was starting the indexing of the array at '1' instead of '0'. This is probably from writing a bunch of Matlab code back in the day (which indexes arrays starting at '1' instead of '0'). Used the suggested for loop and am getting the correct numbers now.

s-ben commented:
@lukebp addressed comments. There were a couple suggestions that could be applied to `manageuser` also (see replies), but wasn't sure if you wanted to do them in this PR.

s-ben commented:
@lukebp have changed hard-coded values in `manageuser.go` to constants. Have a squashed commit ready to push if we're good to merge.
